pkg/sql/catalog/tabledesc/structured_test.go

```diff
@@ -408,28 +450,137 @@ func TestMaybeUpgradeIndexFormatVersion(t *testing.T) {
+						ColumnIDs:   []descpb.ColumnID{1, 2},
+						ColumnNames: []string{"foo", "bar"},
+					},
+				},
+				PrimaryIndex: descpb.IndexDescriptor{
+					ID:                  descpb.IndexID(1),
+					Name:                "primary",
+					KeyColumnIDs:        []descpb.ColumnID{1, 2},
+					KeyColumnNames:      []string{"foo", "bar"},
+					KeyColumnDirections: []descpb.IndexDescriptor_Direction{descpb.IndexDescriptor_ASC, descpb.IndexDescriptor_ASC},
+					EncodingType:        descpb.PrimaryIndexEncoding,
+					Version:             descpb.PrimaryIndexWithStoredColumnsVersion,
+				},
+				Indexes: []descpb.IndexDescriptor{
+					{
+						ID:                  descpb.IndexID(2),
+						Name:                "other",
+						KeyColumnIDs:        []descpb.ColumnID{1},
+						KeyColumnNames:      []string{"foo"},
+						KeyColumnDirections: []descpb.IndexDescriptor_Direction{descpb.IndexDescriptor_ASC},
+						EncodingType:        descpb.SecondaryIndexEncoding,
+						Version:             descpb.StrictIndexColumnIDGuaranteesVersion,
+					},
+					{
+						ID:                  descpb.IndexID(3),
+						Name:                "another",
+						KeyColumnIDs:        []descpb.ColumnID{2},
+						KeyColumnNames:      []string{"bar"},
+						KeyColumnDirections: []descpb.IndexDescriptor_Direction{descpb.IndexDescriptor_ASC},
+						EncodingType:        descpb.SecondaryIndexEncoding,
+						Version:             descpb.StrictIndexColumnIDGuaranteesVersion,
+					},
+				},
+				Privileges: descpb.NewDefaultPrivilegeDescriptor(security.RootUserName()),
+			},
+		},
 	}
 	for i, test := range tests {
-		b := NewBuilder(&test.desc)
-		err := b.RunPostDeserializationChanges(context.Background(), nil)
-		desc := b.BuildImmutableTable()
-		require.NoError(t, err)
-		changes, err := GetPostDeserializationChanges(desc)
-		require.NoError(t, err)
-		err = catalog.ValidateSelf(desc)
-		if test.expValidErr == "" {
+		t.Run(fmt.Sprintf("#%d", i+1), func(t *testing.T) {
+			b := NewBuilder(&test.desc)
+			err := b.RunPostDeserializationChanges(context.Background(), nil)
+			desc := b.BuildImmutableTable()
 			require.NoError(t, err)
-		} else {
-			require.EqualError(t, err, test.expValidErr)
-		}
+			changes, err := GetPostDeserializationChanges(desc)
+			require.NoError(t, err)
+			err = catalog.ValidateSelf(desc)
+			if test.expValidErr != "" {
+				require.EqualError(t, err, test.expValidErr)
+				return
+			}
 
-		upgraded := changes.UpgradedIndexFormatVersion
-		if upgraded != test.expUpgrade {
-			t.Fatalf("%d: expected upgraded=%t, but got upgraded=%t", i, test.expUpgrade, upgraded)
-		}
+			require.NoError(t, err)
+			if test.upgraded == nil {
+				require.Equal(t, PostDeserializationTableDescriptorChanges{}, changes)
+				return
+			}
+
+			if e, a := test.upgraded, desc.TableDesc(); !reflect.DeepEqual(e, a) {
+				for _, diff := range pretty.Diff(e, a) {
+					t.Error(diff)
+				}
+			}
+
+			// Run post-deserialization changes again, descriptor should not change.
```

Comment:
nice

Reply:
Thanks but I wish I'd tested this stuff properly the first time around and saved us a lot of heartache.